Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h

```diff
@@ -656,6 +658,69 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSort(VM& vm, JSGlob
     return JSValue::encode(thisObject);
 }
 
+template<typename ViewClass>
+ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncFromFast(VM& vm, JSGlobalObject* globalObject, CallFrame* callFrame)
+{
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue arrayLike = callFrame->uncheckedArgument(1);
+    JSArrayBufferView* items = jsDynamicCast<JSArrayBufferView*>(arrayLike);
+    if (!items) {
+        // TypedArray.from(Array) case.
+        JSArray* array = jsDynamicCast<JSArray*>(arrayLike);
+        if (!array)
+            return JSValue::encode(jsUndefined());
+
+        if (!array->isIteratorProtocolFastAndNonObservable())
+            return JSValue::encode(jsUndefined());
+
+        IndexingType indexingType = array->indexingType() & IndexingShapeMask;
+        if (indexingType != Int32Shape && indexingType != DoubleShape)
+            return JSValue::encode(jsUndefined());
+
+        size_t length = array->length();
+
+        Structure* structure = globalObject->typedArrayStructure(ViewClass::TypedArrayStorageType);
+        ViewClass* result = ViewClass::createUninitialized(globalObject, structure, length);
+        RETURN_IF_EXCEPTION(scope, { });
+
+        if (indexingType == Int32Shape) {
+            for (unsigned i = 0; i < length; i++) {
+                JSValue value = array->butterfly()->contiguous().at(array, i).get();
+                if (LIKELY(!!value))
+                    result->setIndexQuicklyToNativeValue(i, ViewClass::Adaptor::toNativeFromInt32(value.asInt32()));
+                else
+                    result->setIndexQuickly(i, jsUndefined());
+            }
+        } else {
+            ASSERT(indexingType == DoubleShape);
+            for (unsigned i = 0; i < length; i++) {
+                double d = array->butterfly()->contiguousDouble().at(array, i);
+                if (LIKELY(!std::isnan(d)))
+                    result->setIndexQuicklyToNativeValue(i, ViewClass::Adaptor::toNativeFromDouble(d));
+                else
+                    result->setIndexQuickly(i, jsUndefined());
```

Comment:
Ditto.

Reply:
Ditto here: we know that the source came from a DoubleShape contiguous array.  So, a hole should convert to some constant (though I'm not sure what the spec says it is), right?